src/backend/distributed/commands/multi_copy.c

```diff
@@ -2762,19 +2762,87 @@ ProcessCopyStmt(CopyStmt *copyStatement, char *completionTag, const char *queryS
 				 * COPY table TO PROGRAM / file is handled by wrapping the table
 				 * in a SELECT * FROM table and going through the result COPY logic.
 				 */
-				ColumnRef *allColumns = makeNode(ColumnRef);
 				SelectStmt *selectStmt = makeNode(SelectStmt);
-				ResTarget *selectTarget = makeNode(ResTarget);
 
+#if PG_VERSION_NUM >= PG_VERSION_12
+				List *targetList = NIL;
+				Relation distributedRelation = heap_openrv(copyStatement->relation,
+														   AccessShareLock);
+				TupleDesc tupleDescriptor = RelationGetDescr(distributedRelation);
+				bool hasGeneratedColumns = false;
+
+				for (int i = 0; i < tupleDescriptor->natts; i++)
+				{
+					Form_pg_attribute attr = &tupleDescriptor->attrs[i];
+					if (attr->attisdropped)
+					{
+						continue;
+					}
+
+					if (attr->attgenerated)
+					{
+						hasGeneratedColumns = true;
+						break;
+					}
+				}
+
+				if (hasGeneratedColumns)
+				{
+					for (int i = 0; i < tupleDescriptor->natts; i++)
```

Comment:
I think this function is already big, Maybe we can use functions for this if/else blocks?